src/chain.cc

```diff
@@ -84,29 +85,82 @@ namespace dd
 
+    // actions
+    std::unordered_map<std::string, APIData>::const_iterator ahit
+        = _action_data.begin();
+
+    while (ahit != _action_data.end())
+      {
+        std::string action_id = ahit->first;
+        const APIData &action_data = ahit->second;
+
+        if (action_data.has("output"))
+          {
+            auto out_body = ahit->second.get("output")
+                                .get<oatpp::Any>()
+                                .retrieve<oatpp::Object<DTO::PredictBody>>();
+
+            for (auto p : *out_body->predictions)
+              {
+                std::string uri = p->uri->std_str();
+                p->uri = action_id.c_str();
+                other_models_out.insert(
+                    std::pair<std::string, oatpp::Object<DTO::Prediction>>(uri,
+                                                                           p));
+              }
+          }
+        ++ahit;
+      }
+
     // Return a DTO
     auto chain_dto = DTO::ChainBody::createShared();
     for (auto pred : *first_model_out->predictions)
       {
         oatpp::UnorderedFields<oatpp::Any> chain_pred
             = oatpp_utils::dtoToUFields(pred);
 
+        // chain result at uri level
+        std::string uri = pred->uri->std_str();
+        auto rhit_range = other_models_out.equal_range(uri);
+
+        for (auto rhit = rhit_range.first; rhit != rhit_range.second; ++rhit)
+          {
+            oatpp::String model_name = rhit->second->uri;
+            rhit->second->uri = nullptr;
+            if (chain_pred->find(model_name) != chain_pred->end())
+              throw ChainBadParamException(
+                  "This key already exists and cannot be used to reference "
+                  "model output: "
+                  + model_name->std_str());
+            (*chain_pred)[model_name] = rhit->second;
+          }
+
+        // chain results at prediction level
         auto classes = oatpp::Vector<oatpp::Any>::createShared();
 
         for (auto cls : *pred->classes)
           {
-            std::string uri = cls->class_id->std_str();
-            cls->class_id = nullptr;
             oatpp::UnorderedFields<oatpp::Any> class_preds
                 = oatpp_utils::dtoToUFields(cls);
-            auto rhit_range = other_models_out.equal_range(uri);
 
-            for (auto rhit = rhit_range.first; rhit != rhit_range.second;
-                 ++rhit)
+            if (cls->class_id != nullptr)
               {
-                oatpp::String model_name = rhit->second->uri;
-                rhit->second->uri = nullptr;
-                (*class_preds)[model_name] = rhit->second;
+                uri = cls->class_id->std_str();
+                cls->class_id = nullptr;
+                rhit_range = other_models_out.equal_range(uri);
+
+                for (auto rhit = rhit_range.first; rhit != rhit_range.second;
+                     ++rhit)
+                  {
+                    oatpp::String model_name = rhit->second->uri;
```

Comment:
Make this block code reusable as a function ? it's used before the for loop as well.